src/Appwrite/Platform/Workers/Messaging.php

```diff
@@ -82,85 +85,122 @@ public function action(Message $message, Log $log, Database $dbForProject): void
+        $targetIds = $message->getAttribute('targets', []);
+        $userIds = $message->getAttribute('users', []);
 
         /**
-         * @var Document[] $recipients
+         * @var array<Document> $recipients
          */
         $recipients = [];
 
-        if (\count($topicsId) > 0) {
-            $topics = $dbForProject->find('topics', [Query::equal('$id', $topicsId)]);
+        if (\count($topicIds) > 0) {
+            $topics = $dbForProject->find('topics', [
+                Query::equal('$id', $topicIds),
+                Query::limit(\count($topicIds)),
+            ]);
             foreach ($topics as $topic) {
-                $targets = \array_filter($topic->getAttribute('targets'), fn(Document $target) => $target->getAttribute('providerType') === $message->getAttribute('providerType'));
+                $targets = \array_filter($topic->getAttribute('targets'), fn(Document $target) =>
+                    $target->getAttribute('providerType') === $message->getAttribute('providerType'));
                 $recipients = \array_merge($recipients, $targets);
             }
         }
 
-        if (\count($usersId) > 0) {
-            $users = $dbForProject->find('users', [Query::equal('$id', $usersId)]);
+        if (\count($userIds) > 0) {
+            $users = $dbForProject->find('users', [
+                Query::equal('$id', $userIds),
+                Query::limit(\count($userIds)),
+            ]);
             foreach ($users as $user) {
-                $targets = \array_filter($user->getAttribute('targets'), fn(Document $target) => $target->getAttribute('providerType') === $message->getAttribute('providerType'));
+                $targets = \array_filter($user->getAttribute('targets'), fn(Document $target) =>
+                    $target->getAttribute('providerType') === $message->getAttribute('providerType'));
                 $recipients = \array_merge($recipients, $targets);
             }
         }
 
-        if (\count($targetsId) > 0) {
-            $targets = $dbForProject->find('targets', [Query::equal('$id', $targetsId)]);
+        if (\count($targetIds) > 0) {
+            $targets = $dbForProject->find('targets', [
+                Query::equal('$id', $targetIds),
+                Query::limit(\count($targetIds)),
+            ]);
+            $targets = \array_filter($targets, fn(Document $target) =>
+                $target->getAttribute('providerType') === $message->getAttribute('providerType'));
             $recipients = \array_merge($recipients, $targets);
         }
 
-        $primaryProvider = $dbForProject->findOne('providers', [
+        if (empty($recipients)) {
+            $dbForProject->updateDocument('messages', $message->getId(), $message->setAttributes([
+                'status' => 'failed',
+                'deliveryErrors' => ['No valid recipients found.']
+            ]));
+
+            throw new \Exception('No valid recipients found.');
```

Comment:
I'm not sure we want to throw an exception for these because this error can occur due to a problem the developer did, right? If an exception is thrown, it would get sent to the error logging provider and could eat up our quota, right?

Reply:
Fixed